### PR TITLE
fix(test-app): update sdk to v2.8.0 [ALT-104]

### PR DESCRIPTION
**Pull request by primeinteger**
Updates `test-app` package to the latest SDK

* Runs the SDK in `preview` mode when editing an experience in the UI.
* Runs the SDK in `delivery` mode when accessed outside of the UI (and `?isPreview=true` can be added to the URL to render the draft/preview content outside of the UI)
* Content preview URL can remain as `http://localhost:5173` (no need to add dynamic slug variable)
* Adds rewrite rules for vercel deployment to access pages by slug

### CODE REVIEW

File: packages/test-app/src/Page.tsx [hunk 1]
```diff
@@ -1,51 +1,41 @@
...
 
-  return (
-    <>
-      <ExperienceRoot slug={slug || '/'} experience={experience} locale={'en-US'} />
-    </>
-  );
-};
+  useEffect(() => {
+    if (slug) {
+      fetchBySlug({ experienceTypeId, slug, localeCode });
```
Comment (ryunsong-contentful) [inline packages/test-app/src/Page.tsx:36]: I know this isn't related to your PR, but the fetchBySlug name bothers me as fetching to me always implies some async/await operation. Maybe rename to `getBySlug`?

File: packages/test-app/vercel.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "rewrites": [
```
Comment (ryunsong-contentful) [inline packages/test-app/vercel.json:2]: This is just for my own education - What does this do?

**Review by ryunsong-contentful — APPROVED** [review]
Not too familiar with this repo yet but LGTM

File: packages/test-app/src/Page.tsx (hunk 1 above)
Comment (primeinteger) [reply packages/test-app/src/Page.tsx:36]: The `fetchBySlug` method does an async operation.  I'll add async/await to make it more obvious

File: packages/test-app/vercel.json (hunk 2 above)
Comment (primeinteger) [reply packages/test-app/vercel.json:2]: We have this vercel deployment for the test app package in this repo:
https://vercel.com/contentful-apps/experience-builder-components-test-app

When testing the vercel deployment preview for my dev branch, I noticed any URL with a slug shows a 404:

Screenshot 2023-11-28 at 11 55 38 AM

I looked into it and found this is the default behavior, and you can define a rewrites config that vercel will use to configure the server (https://vercel.com/docs/edge-network/rewrites)

This rewrites config makes so that all page slugs will route to the application endpoint at `/` to match the routing setup we have in [App.tsx](https://github.com/contentful/experience-builder-toolkit/blob/development/packages/test-app/src/App.tsx#L9-L12)

File: packages/test-app/src/Page.tsx [hunk 3]
```diff
@@ -8,7 +8,7 @@ import { createClient } from 'contentful';
 import { useExperienceBuilderComponents } from '@contentful/experience-builder-components';
 import '@contentful/experience-builder-components/styles.css';
 import { useParams } from 'react-router-dom';
-import { useEffect } from 'react';
+import { useEffect, useRef } from 'react';
```
Comment (elylucas) [inline packages/test-app/src/Page.tsx:11]: why is this needed? This kind of set up worries me a bit as people could find this app and think its "best practice" to do something like this, and this is not the type of code we want people to have to write to use EB.

File: packages/test-app/src/Page.tsx [hunk 4]
```diff
@@ -26,17 +26,23 @@ const client = createClient({
 export default function Page() {
   const localeCode = 'en-US';
   const { slug } = useParams<{ slug: string }>();
+  const hasFetched = useRef(false);
```
Comment (elylucas) [inline packages/test-app/src/Page.tsx:29]: why is this needed? This kind of set up worries me a bit as people could find this app and think its "best practice" to do something like this, and this is not the type of code we want people to have to write to use EB.

Comment (elylucas) [inline packages/test-app/src/Page.tsx:29]: why is this needed? This kind of set up worries me a bit as people could find this app and think its "best practice" to do something like this, and this is not the type of code we want people to have to write to use EB.

File: packages/test-app/src/Page.tsx [hunk 5]
```diff
@@ -26,17 +26,23 @@ const client = createClient({
...
   useEffect(() => {
-    (async () => {
+    const asyncFetch = async () => {
       if (slug) {
         await fetchBySlug({ experienceTypeId, slug, localeCode });
       }
-    })();
+    };
+
+    if (!hasFetched.current) {
```
Comment (elylucas) [inline packages/test-app/src/Page.tsx:42]: Ya, not sure this makes sense. This will only ever fetch the experience once, and wouldn't again when the slug changes. So going to a new page wouldn't select a new experience.

File: packages/test-app/src/Page.tsx [hunk 6]
```diff
@@ -26,17 +26,23 @@ const client = createClient({
...
   const { slug } = useParams<{ slug: string }>();
+  const hasFetched = useRef(false);
 
   const { experience, fetchBySlug } = useFetchExperience({ client, mode });
 
   useExperienceBuilderComponents(defineComponents);
 
   useEffect(() => {
-    (async () => {
+    const asyncFetch = async () => {
```
Comment (elylucas) [inline packages/test-app/src/Page.tsx:36]: no need to await on this async call, as `fetchBySlug` will update the experience that got returned with it earlier on its own.

File: packages/test-app/src/Page.tsx (hunk 4 above)
Comment (primeinteger) [reply packages/test-app/src/Page.tsx]: StrictMode makes the useEffect that calls fetchBySlug() to run twice when running the dev server. Adding this ref to track fetch is called makes it only get called once.

The example on the [Wiki page](https://github.com/contentful/experience-builder/wiki#import-and-configuration-of-the-sdk) has this pattern and gave me the impression of being a recommended practice.

I removed the async/await and this ref in the latest commit just now

Comment (elylucas) [reply packages/test-app/src/Page.tsx]: Ok, gotcha. We shouldn't code around dev server idiosyncrasies. They are there to help uncover bugs that might not otherwise show up until production. It also makes the code more convoluted and could cause bugs elsewhere (like the one I pointed out in the other comment). We should take this out and let it double fetch in dev mode, it shouldn't hurt anything and is considered normal practice.

**Review by elylucas — APPROVED** [review]
